### ISSUE TITLE
[Question] why the form of IPO algorithm is not the same as original paper(the log form)

### ISSUE THREAD

**Issue by stvsd1314** [PKU-Alignment/omnisafe#223]
### Required prerequisites

- [X] I have read the documentation <https://omnisafe.readthedocs.io>.
- [X] I have searched the [Issue Tracker](https://github.com/OmniSafeAI/omnisafe/issues) and [Discussions](https://github.com/OmniSafeAI/omnisafe/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [x] Consider asking first in a [Discussion](https://github.com/OmniSafeAI/omnisafe/discussions/new).

### Questions

About the IPO algorithm, why the form is not the same as original paper(the log form)? It makes me confused. And by the way, I find that the effect of using IPO is bad than simply using penalty cost, what is the reason? Thanks !

**Comment by calico-1226** [comment]
In the implementation of the IPO paper, we encountered two issues with the logarithmic form. 
Firstly, because reinforcement learning estimates the sum of costs from sampling to obtain the expected policy, even starting from a feasible policy, the optimization process often encounters situations where the sum of costs exceeds a threshold. This causes the numerical value in the logarithm to be less than 0, making the IPO algorithm unable to run.
Secondly, when the sum of costs is very close to the threshold, the problem of gradient explosion arises due to the pathological curvature in gradient descent caused by the Hessian matrix with a large condition number.

Therefore, we made certain equivalent transformations to the theory of IPO. In situations where the constraints are violated and the gradient is too large, we replaced the gradient with an artificially given maximum value. The specific theoretical analysis is as follows.

In IPO, the loss function is
![196993928-2ff8d83f-0432-43e6-b204-f57974f62bf8](https://user-images.githubusercontent.com/93032279/234579450-b420ecec-fc4b-4b2b-b0aa-f9f9e8ca8885.png)
whose gradient is
![196994159-6e53aed0-17b0-4380-860b-2c19c5027ceb](https://user-images.githubusercontent.com/93032279/234579220-687a29d5-c421-4aea-9c98-c969d9136b67.png)
Thus, we rewrite the loss function as follows,
![196994582-949d6353-c0d6-4988-a9aa-b23012aa55f3](https://user-images.githubusercontent.com/93032279/234579564-f8f5c861-ccc3-4ac8-a48f-bc14d08561c8.png)
where the parameter $\lambda_{i,\theta}$ is defined as the minimum of $\lambda_i(\theta)$ and a maximum value $\lambda_\text{max}$ to avoid issues with ill-conditioning. It is a parameter related to $\theta$ but detached when calculating the gradient flow. Notably, $\mathcal L^\lambda(\theta)$ and $\mathcal L(\theta)$ have the same gradient when $\lambda_{i,\theta}=\lambda_i(\theta)$. Moreover, when IPO violates constraints, we can still update the policy using $L^\lambda(\theta)$ by simply setting $\lambda_{i,\theta}=\lambda_\text{max}$.

**Comment by zmsn-2077** [comment]
Feel free to ask to reopen this if you have more questions.

**Comment by Xi-HHHM** [comment]
Based on the equations mentioned above, the lambda should use the advantage A_c, but the code [here](https://github.com/PKU-Alignment/omnisafe/blob/main/omnisafe/algorithms/on_policy/penalty_function/ipo.py#L68) uses J_c

`penalty = self._cfgs.algo_cfgs.kappa / (self._cfgs.algo_cfgs.cost_limit - Jc + 1e-8)`

Could please explain a little bit why? Thank you!

**Comment by Gaiejj** [comment]
This is an adjustment to the implementation of the IPO algorithm for specific environments. Since the safety cost in Safety-Gymnasium is a non-discounted value with limited steps, the estimation of A_c often cannot accurately reflect the degree of safety violations by the agent. Therefore, we use J_c instead in the code implementation.

**Comment by Andrewllab** [comment]
why to divide (1+penalty) for the [loss calculation](https://github.com/PKU-Alignment/omnisafe/blob/080e6b828c65902611e86e7436dc47c674cdafd7/omnisafe/algorithms/on_policy/penalty_function/ipo.py#L74):
`        return (adv_r - penalty * adv_c) / (1 + penalty)`
Thanks a lot.

**Comment by Xi-HHHM** [comment]
@Andrewllab You can find an explanation here: https://github.com/PKU-Alignment/omnisafe/issues/234

**Comment by Gaiejj** [comment]
Since there has been no response for a long time, we will close this issue. Please feel free to reopen it if you encounter any new problems!
